### PR TITLE
New expressions

### LINKED ISSUE #415: New database functions that R doesn't know about
array_subset (or subset_array)

selected_depth

### PR DESCRIPTION
**Pull request by gergness**
Lots of new expression functions for the R package and make them more directly available to users.

Most of the non-math expressions were hidden behind functions that create Variable Definitions, this made them less composable.

After this PR, the functions that are most commonly used and/or add convenience to the expression get Variable Definition version functions, but we also expose functions that directly create the expression so that complex expressions can be composed. We also expose `crunchdbFunc` which allows users to specify arbitrary zcl functions. I still think it's good to have the expression versions of functions because it's not always obvious how to structure arguments for zcl functions (though this does mean we're duplicating a lot of documentation...)

I put the most effort into a monster documentation file `?expressions`. This became unwieldy when all the methods pointed to it (because R requires that there be a usage entry for each function), so I made a `?expressions-internal` too that holds the usage entries and points to the cleaner one.

Questions:
1) I drew the line that functions that are worth a VarDef version get an integration test, but ones that only have expression generating ones don't. It felt like a lot of new integration tests to me to add them all, but maybe it's worth it.
    > In stand up we said this sounded fine
2) As much as possible, I tried to allow users to avoid specifying category/subvariable ids, allowing them to specify names/aliases. However, when they're working with expressions on top of expressions, there's not much I can do, so ids do poke through. Also, this means there's a lot of ugly code translating between metadata identifiers. Is this the right call?

Work left to do (probably separate issues):
1) `combine_response` and `combine_categories` don't work as expressions of expressions, so I didn't work on them. JJ thought he could fix `combine_categories`, but we'd have to reimplement `combine_responses` into more basic expressions.
2) `interact_variables` and `cut` have R versions that predate expressions. Leaving updating them and adding expression only versions as a separate issue.
3) I believe that updating categories and subvar names of derived variables (with eg `[[<-` should use `alter_categories`) rather than the endpoint as they do now, but I haven't looked into how this would work.
3) We've also now outgrown the print methods for expressions, notably `alterCategories()`. I think I've avoided errors with my last commit, but I may need to dive in.

Closes #415, #447, #41

### CODE REVIEW

**Comment by gergness** [comment]
I think we could use something that combines `case` and `fill`, because they are a lot more useful when you can combine them.

Here's a working version that uses clunky syntax that matches the existing `case` and `fill`.
```r
caseFillExpr <- function(...) {
    case_fills <- list(...)
    
    used_ids <- vapply(case_fills, function(x) x$id %||% NA, numeric(1))
    case_ids <- used_ids
    case_ids[is.na(used_ids)] <- setdiff(seq_along(case_fills), used_ids)[seq_len(sum(is.na(used_ids)))]

    cases <- lapply(seq_along(case_fills), function(cf_num) {
        case_fill <- case_fills[[cf_num]]
        if (!"expression" %in% names(case_fill)) stop("Must have expression")
        
        # Make a temporary cases for expressions that will be filled in
        if ("fill" %in% names(case_fill)) {
            list(
                expression = case_fill$expression, 
                id = as.integer(case_ids[cf_num]), 
                name = paste0("casefill__internal", case_ids[cf_num])
            )
        } else {
            case_fill
        }
    })
    
    need_fills <- vapply(case_fills, function(x) "fill" %in% names(x), logical(1))
    fills <- lapply(which(need_fills), function(cf_num) {
        case_fill <- case_fills[[cf_num]]
        list(fill = case_fill$fill, id = case_ids[cf_num])
    })
    
    fillExpr(caseExpr(cases = cases), fills = fills)
}

ds$x_casefill <- caseFillExpr(
    list(expression = ds$x %in% c("a", "b"), fill = ds$y),
    list(expression = ds$x %in% c("c", "d"), name = "c or d"),
    list(expression = "else", name = "xxx")
)
```

I actually think this syntax is better, and think it might be worth the inconsistency. (modelled on `dplyr::case_when()`.

```r
caseWhenExpr(
    ds$x %in% c("a", "b") ~ ds$y,
    ds$x %in% c("c", "d") ~ Category(name = "c or d"),
    TRUE ~ Category(name = "xxx")
)
```

**Comment by gergness** [comment]
Documentation tweaks that have come up today from Matt:

`VariableCatalog` documentation should point to `allVariables()` and `variables()` and maybe scan other class documentation to point to a way to actually get one (`CrunchDataset`?, others?)


```
also… with straightlineResponse() the documentation says it considered the “responses”… but perhaps that language is too ambiguous ?

Perhaps we need to be clear it’s the categories or the values or the order ?
```

**Comment by gergness** [comment]
I think I also need to split out documentation for the expression functions. Matt was able to successfully find `arraySubsetExpr`, but couldn't get it to work

File: tests/testthat/test-rowwise.R [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                name = "x"
+            )
+        )
+    })
+
+    test_that("error for non-array var straightline", {
+        expect_error(straightlineResponse(ds$textvar), "x must be an array variable")
+    })
+
+    test_that("tieredVar creates correct vardef", {
```
Comment (malecki) [inline tests/testthat/test-rowwise.R:42]: remove tests if they are more confusing to read and don’t really seem to add much. (does R really need to claim understanding of 'tiered'?)

File: NAMESPACE [hunk 2]
```diff
@@ -238,6 +258,7 @@ export(newMultitable)
 export(newProject)
 export(newSlide)
 export(noTransforms)
+export(normalize)
```
[the remark this answers is not in the document]
Comment (gergness) [reply NAMESPACE:261]: renamed to `crunchNormalize` - my recollection of our conversation was just that the name was too broad

**Pull request merged**